### PR TITLE
[5.1] Joomla Update : extensions check never ends

### PR DESCRIPTION
**Pull request by conseilgouz**
Pull Request for Issue #43196.

### Summary of Changes

Fix PHP 8.2 deprecated notices during Extensions check while performing Joomla 5.1.0 update

### Testing Instructions
Joomla 5.0.3, PHP 8.2 or higher, enable Error reporting to default.
In Joomla Update configuration, set "Potentially incompatible extensions checkbox" to "Show".
Install any non core extension

Update Joomla to 5.1.0-RC shows the Pre-Update Screen.

### Actual result BEFORE applying this Pull Request

Extensions check never ends, an error appears in explorer console : uncaught syntaxError.. 

### Expected result AFTER applying this Pull Request

Extensions check works fine

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [X] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [X] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by conseilgouz** [comment]
Administrator/components/com_joomlaupdate\src\Model\UpdateModel.php creates a new Update object in function checkCompatibility, line 1629 and below. It tries to set properties that do not exist in Update object.
This fix creates the missing properties.
Note : $jversion.full has been renamed $jversionfull.

**Comment by brianteeman** [comment]
I still cannot replicate the reported error. Maybe I am missing a step. Can you please make a screen recording showing everything you did to create the error. I use licecap for this.

**Comment by conseilgouz** [comment]
https://github.com/joomla/joomla-cms/assets/19435246/83293025-23b8-4a90-a4f3-dad62eb48c03

**Comment by richard67** [comment]
Does the issue also happen with error reporting = none?
If no, then we have a readable workaround and we just should recommend that in the release announcement.

And does it also happen on 4.4? If yes, the PR should be made for the 4.4-dev branch, and that will later be merged up into 5.1-dev by maintainers. If no, the PR should be made for the 5.1-dev branch as there will not be any new 5.0.x release.

**Comment by brianteeman** [comment]
finaly I can replicate it. I was performing the update using the upload and update function and that produces no errors

**Comment by conseilgouz** [comment]
I just tried a Joomla 4.4.3 environment and this issue does not occur. 
Joomla 5 update class uses LegacyPropertyManagementTrait which causes notices display.

**Comment by brianteeman** [comment]
LegacyPropertyManagementTrait  was introduced by @laoneo https://github.com/joomla/joomla-cms/pull/40999 for 5.0.0

**Comment by conseilgouz** [comment]
Oups, I blow it up while trying to rebase it to Joomla 5.1

**Comment by richard67** [comment]
@conseilgouz Have you just tried to rebase your branch? If so, it seems there went something Wrong. You should change the base branch on GitHub by using the "Edit" button right beside the PR title. Let me know if I shall do that for you.

**Comment by conseilgouz** [comment]
@richard67 : well, after trying rebase from my PC, 1120 updated files ???? your help is more than welcomed.

**Comment by richard67** [comment]
> @richard67 : well, after trying rebase from my PC, 1120 updated files ???? your help is more than welcomed.

@conseilgouz As said, it needs to change the base branch on GitHub. I will do that now.

**Comment by conseilgouz** [comment]
what about the 1120 updated files ?

**Comment by richard67** [comment]
> what about the 1120 updated files ?

@conseilgouz Check again. Now there are only 4 remaining. But there are still some unrelated changes in files `plugins/content/vote/src/Extension/Vote.php` and `plugins/content/vote/src/Extension/Vote.php` shown. Could you check and revert?

**Comment by conseilgouz** [comment]
it still contains my other PR#42933

**Comment by richard67** [comment]
> it still contains my other PR#42933

@conseilgouz Yes, because you have rebased to YOUR 5.1-dev branch, and you have made the other PR not on a separate branch but on the 5.1-dev branch. You should have rebased this PR to the UPSTREAM 5.1-dev.

**Comment by richard67** [comment]
@conseilgouz P.S.: Just revert these changes here and this PR here is ok.

File: libraries/src/Updater/Update.php [hunk 1]
```diff
@@ -256,6 +256,10 @@ class Update
     protected $stability;
     protected $supported_databases;
     protected $php_minimum;
+    protected $section;
```
Comment (Quy) [inline libraries/src/Updater/Update.php]: Duplicate. See line 266.

Comment (conseilgouz) [reply libraries/src/Updater/Update.php]: Fixed

**Comment by brianteeman** [comment]
are you sure that it is ok to rename `jversion.full` to `jversionfull` ?

**Comment by conseilgouz** [comment]
> are you sure that it is ok to rename `jversion.full` to `jversionfull` ?

I checked in all the sources and it's only used in Update class and in JoomlaUpdate Model.

**Comment by brianteeman** [comment]
but why rename it?

**Comment by conseilgouz** [comment]
I did not find a way to define a variable with this kind of name (with a dot).

**Comment by brianteeman** [comment]
just seems wrong that to fix the bug you need to change the name of something that has been in use for a very long time. 

@laoneo as you made the change to LegacyPropertyManagementTrait in https://github.com/joomla/joomla-cms/pull/40999 for 5.0.0 does this look correct to you.

It smells of being a bandaid to fix a bigger issue

**Comment by conseilgouz** [comment]
jversion.full is just a variable.

It's only used in Update class with a preg_match (line 364) : preg_match('/^' . $this->currentUpdate->targetplatform->version . '/', $this->get('jversion.full', JVERSION))

In PHP 8.2, we have to define it before using it in an object, otherwise it sends a deprecated notice but it creates it anyway.
In PHP 9, this will cause an error.

I had to change variable name as I did not find a way of declaring a variable with a dot in its name.

FYI, in Joomla 3.10, jversion.full was already there and there was also jversion.dev_level (never initialized, so containing Version::PATCH_VERSION). Both variables were used in Update class the same way they were in later Joomla versions.

**Comment by bembelimen** [comment]
Thanks @conseilgouz for the finding. We have to discuss how we handle this issue best. It need to be fixed asap, but either it will trigger a new 5.0.x or we will move it to 5.1.1. Sadly 5.1.0 is the wrong version for it, but we get it managed :)

Thanks for the finding and the PR.

**Comment by laoneo** [comment]
@conseilgouz can you rename the variable `jversionfull` to $targetVersion? Like that it is more meaningful. Also remove the set and get calls and make a proper setter function like `setTargetVersion for it with the? Thanks...

File: libraries/src/Updater/Update.php [hunk 2]
```diff
@@ -372,7 +382,7 @@ public function _endElement($parser, $name)
                 if (
                     isset($this->currentUpdate->targetplatform->name)
                     && $product == $this->currentUpdate->targetplatform->name
-                    && preg_match('/^' . $this->currentUpdate->targetplatform->version . '/', $this->get('jversion.full', JVERSION))
+                    && preg_match('/^' . $this->currentUpdate->targetplatform->version . '/',$this->getTargetVersion())
```
Comment (richard67) [inline libraries/src/Updater/Update.php]:
```suggestion
                    && preg_match('/^' . $this->currentUpdate->targetplatform->version . '/', $this->getTargetVersion())
```
Fix PHPCS

Comment (conseilgouz) [reply libraries/src/Updater/Update.php]: done

File: libraries/src/Updater/Update.php [hunk 3]
```diff
@@ -690,4 +700,36 @@ protected function stabilityTagToInteger($tag)
...
+     */
+    public function getTargetVersion()
+    {
+        if (!$this->targetVersion) {
+            return JVERSION;
+        }
+        return $this->targetVersion;
+    }
+
 }
```
Comment (richard67) [inline libraries/src/Updater/Update.php]:
```suggestion
    }
}
```
Now after the first PHPCS tool is happy, the 2nd (and last) one has found this. If this is fixed, PHPCS should be successful in Drone.

File: libraries/src/Updater/Update.php [hunk 4]
```diff
@@ -256,11 +256,21 @@ class Update
...
     public $sha512;
     protected $section;
 
+    /**
+     * Joomla! target version used by the pre-update check
+     *
+     * @var    string
+     * @since  __DEPLOY_VERSION__
+     */
+    protected $targetVersion;
```
Comment (laoneo) [inline libraries/src/Updater/Update.php]: This should be private.

File: libraries/src/Updater/Update.php [hunk 5]
```diff
@@ -690,4 +700,36 @@ protected function stabilityTagToInteger($tag)
...
+     */
+    public function setTargetVersion($version)
+    {
+        $this->targetVersion = $version;
+    }
+
+    /**
+     * Get extension's Joomla! target version
+     *
+     * @param   string  $version  The target version
```
Comment (laoneo) [inline libraries/src/Updater/Update.php]: There is no parameter here

File: libraries/src/Updater/Update.php [hunk 6]
```diff
@@ -256,11 +256,21 @@ class Update
     protected $stability;
     protected $supported_databases;
     protected $php_minimum;
+    protected $folder;
```
Comment (laoneo) [inline libraries/src/Updater/Update.php:259]: Is $folder and $changelogurl used outside of the class?

**Comment by laoneo** [comment]
> @laoneo : ok for targetVersion. In progress... What about other variables that have been added "the hard way" : downloadurl, tag, stability... ? It's not the purpose of this PR, but could be cleaner

You mean to make getter and setter functions for them?

File: libraries/src/Updater/Update.php (hunk 6 above)
Comment (conseilgouz) [reply libraries/src/Updater/Update.php:259]: Well, to be frank, I don't know.
They are created in _endElement function using  a foreach (get_object_vars($this->latest) .... and I did not find  where folder and changelogurl are being used.

File: libraries/src/Updater/Update.php (hunk 5 above)
Comment (conseilgouz) [reply libraries/src/Updater/Update.php]: done

**Comment by conseilgouz** [comment]
> > @laoneo : ok for targetVersion. In progress... What about other variables that have been added "the hard way" : downloadurl, tag, stability... ? It's not the purpose of this PR, but could be cleaner
> 
> You mean to make getter and setter functions for them?

Well, I was thinking about this, but it's tricky to find the "getters" and the variables are set  in _endElement thru a nice                     foreach (get_object_vars($this->latest) as $key => $val) {
                        $this->$key = $val;
                    }
So the setters are also not that easy to create.

**Review by laoneo — APPROVED** [review]

File: libraries/src/Updater/Update.php [hunk 7]
```diff
@@ -690,4 +700,33 @@ protected function stabilityTagToInteger($tag)
...
+     * @return  string
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function getTargetVersion()
+    {
+        if (!$this->targetVersion) {
+            return JVERSION;
+        }
+        return $this->targetVersion;
```
Comment (richard67) [inline libraries/src/Updater/Update.php:731]:
```suggestion

        return $this->targetVersion;
```
One last code style thing.

**Comment by conseilgouz** [comment]
Everything seems to be ok (thank you @richard67, @laoneo for the time you spent ), but, it applies to Joomla 5.1 when initial issue was about "J5.0.3 update to J5.1.0 does not work".
Did it worth the time we spent ?

**Comment by richard67** [comment]
@conseilgouz Well we can’t fix it for 5.0.3 with a CMS update, and we have a workaround for this, switching off error reporting, but we need the fix also for later updates from 5.1.0 to e.g. 5.1.1 in future, so it is ok for 5.1.

**Comment by richard67** [comment]
@conseilgouz You‘ve deleted your fork of the joomla repository, so this PR was closed. Was this a mistake or by purpose?

**Comment by conseilgouz** [comment]
Hi Richard,
I had 2 other PR's which said I could delete my fork and I forgot about this one.
I thought it was already closed.
need a "refork" ?

**Comment by richard67** [comment]
> need a "refork" ?

@conseilgouz I think so. The PR was not closed by us, and it was approved by @laoneo , so it would have just needed 2 human tests. If you still want to provide a PR, I think you have to fork again and redo the PR.

**Comment by richard67** [comment]
New PR is #43410 . @laoneo Could you approve that one, too? It’s the same as this one here which had been closed by mistake.

**Pull request closed without merging**